gateway-sp-comms/src/management_switch/location_map.rs

```diff
@@ -0,0 +1,461 @@
+    for (port, config) in port_config {
+        futs.push(async move {
+            // construct a socket pointed to a multicast addr instead of a
+            // specific, known addr
+            let socket = SpSocket {
+                port,
+                addr: config.multicast_addr,
+                // all ports in `port_config` also get sockets bound to them;
+                // unwrapping this lookup is fine
+                socket: sockets.get(&port).unwrap(),
+            };
+
+            let mut backoff = backoff::internal_service_policy();
+            loop {
+                let duration = backoff
+                    .next_backoff()
+                    .expect("internal backoff policy gave up");
+                tokio::time::sleep(duration).await;
+
+                let result = communicator
+                    .request_response(
+                        &socket,
+                        RequestKind::Discover,
+                        ResponseKindExt::try_into_discover,
+                        // TODO should this timeout be configurable or itself
+                        // have some kind of backoff? we're inside a
+                        // `backoff::retry()` loop, but if an SP is alive but
+                        // slow (i.e., taking longer than this timeout to reply)
+                        // we'll never hear it - the response will show up late
+                        // and we'll ignore it. For now just leave it at some
+                        // reasonably large number; this may solve itself when
+                        // we move to some kind of authenticated comms channel.
+                        Some(Instant::now() + Duration::from_secs(5)),
+                    )
+                    .await;
+
+                match result {
+                    Ok(response) => return (port, response),
+                    Err(err) => {
+                        debug!(
+                            log, "discovery failed; will retry";
+                            "port" => ?port,
+                            "err" => %err,
+                        );
+                    }
+                }
+            }
+        });
+    }
+
+    // Wait for responses.
+    while let Some((port, response)) = futs.next().await {
+        // See if this port can participate in location determination.
+        let pos = match location_determination
+            .iter()
+            .position(|d| d.switch_port == port.0)
+        {
+            Some(pos) => pos,
+            None => {
+                info!(
+                    log, "received discovery response (not used for location)";
+                    "port" => ?port,
+                    "response" => ?response,
+                );
+                continue;
+            }
+        };
+        let determination = location_determination.remove(pos);
+
+        let refined = match response.sp_port {
+            0 => determination.sp_port_0,
+            1 => determination.sp_port_1,
```

Comment:
do I recall correctly that the KSZ numbers these 1 and 2? If that's the case we might want to stick with that to prevent confusion in the future.

Reply:
Fixed in 7782fbc